File: web-api/src/runtime.ts
import { spawn } from 'node:child_process'
import fs from 'node:fs/promises'
import path from 'node:path'
import readline from 'node:readline'
import type { ChildProcess } from 'node:child_process'
import type { RuntimeHealth, StartJobRequest, WebApiConfig } from './types.js'

const IDENTITY_ENCODING_DRIVER_SOURCE = `export default async ({ page, data, crawler, seed }) => {
  await page.setExtraHTTPHeaders({ "Accept-Encoding": "identity" });
  await crawler.loadPage(page, data, seed);
};
`
const ZIMIT_CONVERSION_HEADROOM_SECONDS = 15 * 60
const ATTEMPT_TIMEOUT_HEADROOM_SECONDS = 60

export interface DockerRunResult {
  success: boolean
  errorMessage?: string
  retryable?: boolean
  exitCode?: number | null
}

export interface ZimitRunOptions {
  resumeConfigPath?: string | null
  saveStateIntervalSeconds?: number
}

const dockerEnv = (config: WebApiConfig): NodeJS.ProcessEnv => {
  const env: NodeJS.ProcessEnv = { ...process.env }

  if (config.dockerHost) {
    env.DOCKER_HOST = config.dockerHost
  } else if (config.dockerSocketPath) {
    env.DOCKER_HOST = `unix://${config.dockerSocketPath}`
  }

  return env
}

const commandResult = (
  command: string,
  args: string[],
  config: WebApiConfig,
): Promise<{ success: boolean; output: string }> =>
  new Promise((resolve, reject) => {
    const child = spawn(command, args, {
      env: dockerEnv(config),
      stdio: ['ignore', 'pipe', 'pipe'],
    })

    let output = ''
    child.stdout.on('data', (chunk) => {
      output += chunk.toString()
    })
    child.stderr.on('data', (chunk) => {
      output += chunk.toString()
    })
    child.on('error', (error) => reject(error))
    child.on('close', (code) => {
      resolve({
        success: code === 0,
        output,
      })
    })
  })

const commandSuceeds = async (
  command: string,
  args: string[],
  config: WebApiConfig,
): Promise<boolean> => {
  try {
    const result = await commandResult(command, args, config)
    return result.success
  } catch {
    return false
  }
}

const isDockerNoiseLine = (line: string): boolean => {
  const lowered = line.toLowerCase()
  if (lowered.startsWith('digest:') || lowered.startsWith('status:')) {
    return true
  }

  if (lowered.includes('pulling from') || lowered.includes('pull complete')) {
    return true
  }

  const parts = line.split(':')
  if (parts.length >= 2) {
    const layerPrefix = parts[0]
    const isHexLayer = layerPrefix.length >= 12 && /^[0-9a-f]+$/i.test(layerPrefix)
    if (isHexLayer) {
      const suffix = parts.slice(1).join(':').trim().toLowerCase()
      return (
        suffix === 'already exists' ||
        suffix === 'download complete' ||
        suffix === 'pull complete' ||
        suffix === 'waiting' ||
        suffix.startsWith('extracting') ||
        suffix.startsWith('verifying checksum')
      )
    }
  }

  return false
}

const normalizeZimitLogLine = (line: string): string | null => {
  try {
    const parsed = JSON.parse(line) as {
      message?: string
      context?: string
      logLevel?: string
      details?: Record<string, unknown>
    }

    const message = parsed.message || ''
    const context = parsed.context || ''
    const level = parsed.logLevel || ''
    const details = parsed.details || {}

    if (context === 'redis' && level === 'warn') {
      return null
    }

    if (message === 'Starting page') {
      const page = typeof details.page === 'string' ? details.page : 'unknown'
      return `Starting page crawl: ${page}`
    }
    if (message === 'Page Finished') {
      const page = typeof details.page === 'string' ? details.page : 'unknown'
      return `Page finished: ${page}`
    }
    if (message === 'Crawl statistics') {
      const crawled = Number(details.crawled ?? 0)
      const total = Number(details.total ?? 0)
      const pending = Number(details.pending ?? 0)
      const failed = Number(details.failed ?? 0)
      return `Crawl progress: ${crawled}/${total} crawled, ${pending} pending, ${failed} failed`
    }
    if (message === 'Crawling done') {
      return 'Crawl completed. Starting ZIM conversion...'
    }
    if (!message) {
      return null
    }

    if (!context || ['general', 'worker', 'pageStatus', 'crawlStatus'].includes(context)) {
      return message
    }

    return `${context}: ${message}`
  } catch {
    return line
  }
}

const pushRecentLine = (recentLines: string[], line: string): void => {
  const trimmed = line.trim()
  if (!trimmed) {
    return
  }

  const compact =
    trimmed.length > 280 ? `${trimmed.slice(0, 280).trimEnd()}...` : trimmed
  if (recentLines[recentLines.length - 1] === compact) {
    return
  }
  recentLines.push(compact)
  if (recentLines.length > 6) {
    recentLines.shift()
  }
}

const dockerMountPath = (value: string): string => {
  if (/^[A-Za-z]:\\/.test(value)) {
    const drive = value[0].toLowerCase()
    const rest = value.slice(2).replace(/\\/g, '/').replace(/^\/+/, '')
    return `/${drive}/${rest}`
  }
  return value
}

const hostPathToOutputContainerPath = (
  hostPath: string,
  outputDirectory: string,
): string | null => {
  const relativePath = path.relative(outputDirectory, hostPath)
  if (
    relativePath.length === 0 ||
    relativePath === '.' ||
    (!relativePath.startsWith('..') && !path.isAbsolute(relativePath))
  ) {
    const normalized =
      relativePath.length === 0 || relativePath === '.'
        ? ''
        : relativePath.replace(/\\/g, '/')
    return normalized.length > 0 ? `/output/${normalized}` : '/output'
  }
  return null
}

const makeDriverPath = (outputDirectory: string, containerName: string): string =>
  path.join(outputDirectory, `.zimple-driver-${containerName}.mjs`)

const filesystemUsagePercent = (targetPath: string): Promise<number | null> =>
  new Promise((resolve) => {
    const child = spawn('df', ['-P', targetPath], {
      stdio: ['ignore', 'pipe', 'ignore'],
    })
    let output = ''

    child.stdout.on('data', (chunk) => {
      output += chunk.toString()
    })
    child.on('error', () => {
      resolve(null)
    })
    child.on('close', (code) => {
      if (code !== 0) {
        resolve(null)
        return
      }

      const lines = output
        .split('\n')
        .map((line) => line.trim())
        .filter(Boolean)
      const usageLine = lines[1]
      if (!usageLine) {
        resolve(null)
        return
      }

      const percentField = usageLine
        .split(/\s+/)
        .find((value) => value.endsWith('%'))
      if (!percentField) {
        resolve(null)
        return
      }

      const percent = Number.parseInt(percentField.slice(0, -1), 10)
      resolve(Number.isNaN(percent) ? null : percent)
    })
  })

const escapeRegexFragment = (value: string): string =>
  value.replace(/[\\^$.*+?()[\]{}|/-]/g, '\\$&')

const normalizeScopedPath = (pathname: string): string | null => {
  if (!pathname || pathname === '/') {
    return null
  }

  const trimmed = pathname.replace(/\/+$/, '')
  if (!trimmed || trimmed === '/') {
    return null
  }

  return trimmed
}

const derivePathPrefixIncludePattern = (targetUrl: string): string | null => {
  try {
    const parsed = new URL(targetUrl)
    const scopedPath = normalizeScopedPath(parsed.pathname)
    if (!scopedPath) {
      return null
    }
    return `^https?://[^/]+${escapeRegexFragment(scopedPath)}(?:$|[/?#].*)`
  } catch {
    return null
  }
}

const effectiveIncludePatterns = (request: StartJobRequest): string[] => {
  if (request.crawl.includePatterns.length > 0) {
    return request.crawl.includePatterns
  }

  const derived = derivePathPrefixIncludePattern(request.url)
  return derived ? [derived] : []
}

export const containerNameForJob = (jobId: string): string => {
  const suffix = jobId.slice(0, 12)
  return `zimple-${suffix}`
}

export const retryDelaySeconds = (attemptIndex: number): number => {
  const exponent = Math.min(Math.max(attemptIndex, 0), 4)
  return 2 ** exponent
}

export const sleepForRetry = (attemptIndex: number): Promise<void> =>
  new Promise((resolve) => {
    setTimeout(resolve, retryDelaySeconds(attemptIndex) * 1000)
  })

export const zimitTimeHardLimitSeconds = (timeoutMinutes: number): number =>
  Math.max(timeoutMinutes, 1) * 60 + ZIMIT_CONVERSION_HEADROOM_SECONDS

export const zimitAttemptTimeoutMs = (timeoutMinutes: number): number =>
  (zimitTimeHardLimitSeconds(timeoutMinutes) + ATTEMPT_TIMEOUT_HEADROOM_SECONDS) * 1000

export const buildDockerArgs = (
  request: StartJobRequest,
  outputDirectory: string,
  outputFilename: string,
  containerName: string,
  zimitImage: string,
  driverContainerPath?: string,
  options?: ZimitRunOptions,
): string[] => {
  const includePatterns = effectiveIncludePatterns(request)
  const sizeHardLimitBytes = Math.max(request.crawl.limits.maxTotalSizeMb, 1) * 1024 * 1024
  const timeHardLimitSeconds = zimitTimeHardLimitSeconds(request.crawl.limits.timeoutMinutes)
  const saveStateIntervalSeconds = Math.max(
    Math.floor(options?.saveStateIntervalSeconds ?? 60),
    10,
  )
  const args = [
    'run',
    '--rm',
    '--name',
    containerName,
    '-v',
    `${dockerMountPath(outputDirectory)}:/output`,
    zimitImage,
    'zimit',
    '--seeds',
    request.url.trim(),
    '--name',
    outputFilename,
    '--output',
    '/output',
    '--scopeType',
    includePatterns.length > 0 ? 'custom' : 'domain',
    '--diskUtilization',
    '0',
    '-w',
    String(request.crawl.workers),
    '--depth',
    String(request.crawl.limits.maxDepth),
    '--pageLimit',
    String(request.crawl.limits.maxPages),
    '--maxPageRetries',
    String(request.crawl.limits.retries),
    '--timeHardLimit',
    String(timeHardLimitSeconds),
    '--sizeHardLimit',
    String(sizeHardLimitBytes),
    '--saveState',
    'partial',
    '--saveStateInterval',
    String(saveStateIntervalSeconds),
    '--keep',
  ]

  if (driverContainerPath) {
    args.push('--driver', driverContainerPath)
  }

  if (options?.resumeConfigPath) {
    const resumeConfigContainerPath = hostPathToOutputContainerPath(
      options.resumeConfigPath,
      outputDirectory,
    )
    if (resumeConfigContainerPath) {
      args.push('--config', resumeConfigContainerPath)
    }
  }

  for (const pattern of includePatterns) {
    args.push('--scopeIncludeRx', pattern)
  }
  for (const pattern of request.crawl.excludePatterns) {
    args.push('--scopeExcludeRx', pattern)
  }

  return args
}

export const checkRuntimeHealth = async (config: WebApiConfig): Promise<RuntimeHealth> => {
  const dockerInstalled = await commandSuceeds('docker', ['--version'], config)
  if (!dockerInstalled) {
    return {
      dockerInstalled: false,
      dockerResponsive: false,
      zimitImagePresent: false,
      ready: false,
      message: 'Docker is not installed or is not available on PATH.',
    }
  }

  const dockerResponsive = await commandSuceeds(
    'docker',
    ['info', '--format', '{{.ServerVersion}}'],
    config,
  )
  if (!dockerResponsive) {
    return {
      dockerInstalled: true,
      dockerResponsive: false,
      zimitImagePresent: false,
      ready: false,
      message: 'Docker is installed but the daemon is not reachable.',
    }
  }

  const zimitImagePresent = await commandSuceeds(
    'docker',
    ['image', 'inspect', config.zimitImage],
    config,
  )

  return {
    dockerInstalled: true,
    dockerResponsive: true,
    zimitImagePresent,
    ready: true,
    message: zimitImagePresent
      ? 'Runtime ready. ZIM jobs can be started.'
      : `Docker is ready. Pulling ${config.zimitImage} may be required on first run.`,
  }
}

export const ensureZimitImage = async (
  config: WebApiConfig,
  timeoutMs: number,
): Promise<boolean> => {
  const inspect = await commandResult(
    'docker',
    ['image', 'inspect', config.zimitImage],
    config,
  )
  if (inspect.success) {
    return false
  }

  const pullChild = spawn('docker', ['pull', '--quiet', config.zimitImage], {
    env: dockerEnv(config),
    stdio: ['ignore', 'ignore', 'pipe'],
  })

  const result = await Promise.race([
    new Promise<{ code: number | null; stderr: string }>((resolve, reject) => {
      let stderr = ''
      pullChild.stderr.on('data', (chunk) => {
        stderr += chunk.toString()
      })
      pullChild.on('error', reject)
      pullChild.on('close', (code) => resolve({ code, stderr }))
    }),
    new Promise<{ code: number | null; stderr: string }>((resolve) => {
      setTimeout(() => {
        pullChild.kill('SIGTERM')
        resolve({
          code: null,
          stderr: 'Timed out while preparing zimit runtime image.',
        })
      }, timeoutMs)
    }),
  ])

  if (result.code === 0) {
    return true
  }

  const lastError = result.stderr
    .split('\n')
    .map((line) => line.trim())
    .filter(Boolean)
    .at(-1)

  throw new Error(
    `Failed to prepare zimit runtime image: ${lastError || 'Unknown Docker pull error'}`,
  )
}

export const runZimitOnce = async (
  config: WebApiConfig,
  request: StartJobRequest,
  outputDirectory: string,
  outputFilename: string,
  containerName: string,
  onLog: (line: string) => void,
  onProcess: (child: ChildProcess) => void,
  options?: ZimitRunOptions,
): Promise<DockerRunResult> => {
  await fs.mkdir(outputDirectory, { recursive: true })

  const driverPath = makeDriverPath(outputDirectory, containerName)
  await fs.writeFile(driverPath, IDENTITY_ENCODING_DRIVER_SOURCE, 'utf8')

  const args = buildDockerArgs(
    request,
    outputDirectory,
    outputFilename,
    containerName,
    config.zimitImage,
    `/output/${path.basename(driverPath)}`,
    options,
  )

  const child = spawn('docker', args, {
    env: dockerEnv(config),
    stdio: ['ignore', 'pipe', 'pipe'],
  })
  onProcess(child)

  const recentOutput: string[] = []
  const consume = (stream: NodeJS.ReadableStream): Promise<void> =>
    new Promise((resolve) => {
      const reader = readline.createInterface({ input: stream })
      reader.on('line', (line) => {
        if (isDockerNoiseLine(line)) {
          return
        }

        const normalized = normalizeZimitLogLine(line)
        if (!normalized) {
          return
        }
        pushRecentLine(recentOutput, normalized)
        onLog(normalized)
      })
      reader.on('close', resolve)
    })

  const waitForExit = new Promise<number | null>((resolve, reject) => {
    child.on('error', reject)
    child.on('close', (code) => resolve(code))
  })

  try {
    await Promise.all([consume(child.stdout), consume(child.stderr), waitForExit])
    const exitCode = await waitForExit

    if (exitCode === 0) {
      return { success: true, retryable: false, exitCode }
    }

    let message =
      exitCode === null
        ? 'zimit container terminated by signal.'
        : `zimit container failed with exit code ${exitCode}.`
    let retryable = true

    if (exitCode === 2) {
      retryable = false
      message +=
        ' Exit code 2 usually means zimit rejected one or more options or the output archive name already exists.'
    }
    if (exitCode === 3) {
      retryable = false
      const usage = await filesystemUsagePercent(outputDirectory)
      message +=
        ' Exit code 3 indicates browsertrix stopped due output filesystem space/utilization checks. Browsertrix has a hard stop when output filesystem usage reaches 99% or more.'
      if (usage !== null) {
        message += ` Detected output filesystem usage: ${usage}%.`
      }
    }
    if (exitCode === 15) {
      retryable = false
      if (recentOutput.some((line) => line.includes('Crawl completed. Starting ZIM conversion'))) {
        message +=
          ' Exit code 15 after crawl completion usually means the runtime was interrupted during conversion by a time limit; increase timeout minutes.'
      } else {
        message +=
          ' Exit code 15 indicates the runtime was interrupted (timeout or external stop request).'
      }
    }
    if (recentOutput.length > 0) {
      message += ` Last output: ${recentOutput.join(' | ')}`
    }

    return {
      success: false,
      errorMessage: message,
      retryable,
      exitCode,
    }
  } finally {
    await fs.unlink(driverPath).catch(() => undefined)
  }
}

export const stopContainer = async (
  config: WebApiConfig,
  containerName: string,
): Promise<boolean> =>
  commandSuceeds('docker', ['rm', '-f', containerName], config)

File: src-tauri/src/runtime.rs
use crate::models::{RuntimeHealth, StartJobRequest};
use serde_json::Value;
use std::collections::VecDeque;
use std::fs;
use std::path::{Path, PathBuf};
use std::process::Stdio;
use std::time::Duration;
use tokio::io::{AsyncBufReadExt, BufReader};
use tokio::process::Command;
use tokio::sync::mpsc;

const ZIMIT_IMAGE: &str = "ghcr.io/openzim/zimit";
const IDENTITY_ENCODING_DRIVER_SOURCE: &str = r#"export default async ({ page, data, crawler, seed }) => {
  await page.setExtraHTTPHeaders({ "Accept-Encoding": "identity" });
  await crawler.loadPage(page, data, seed);
};
"#;

#[derive(Debug, Clone)]
pub struct DockerRunResult {
    pub success: bool,
    pub error_message: Option<String>,
}

fn extract_docker_error(stderr: &str) -> String {
    stderr
        .lines()
        .rev()
        .map(str::trim)
        .find(|line| !line.is_empty())
        .map(ToString::to_string)
        .unwrap_or_else(|| "Unknown Docker error".to_string())
}

fn is_docker_noise_line(line: &str) -> bool {
    let lowered = line.to_ascii_lowercase();
    if lowered.starts_with("digest:") || lowered.starts_with("status:") {
        return true;
    }

    if lowered.contains("pulling from") || lowered.contains("pull complete") {
        return true;
    }

    if let Some((prefix, suffix)) = line.split_once(':') {
        let is_hex_layer = prefix.len() >= 12 && prefix.chars().all(|ch| ch.is_ascii_hexdigit());
        if is_hex_layer {
            let suffix = suffix.trim().to_ascii_lowercase();
            return suffix == "already exists"
                || suffix == "download complete"
                || suffix == "pull complete"
                || suffix == "waiting"
                || suffix.starts_with("extracting")
                || suffix.starts_with("verifying checksum");
        }
    }

    false
}

fn normalize_zimit_log_line(line: &str) -> Option<String> {
    let parsed: Value = match serde_json::from_str(line) {
        Ok(value) => value,
        Err(_) => return Some(line.to_string()),
    };

    let message = parsed
        .get("message")
        .and_then(Value::as_str)
        .unwrap_or_default();
    let context = parsed
        .get("context")
        .and_then(Value::as_str)
        .unwrap_or_default();
    let level = parsed
        .get("logLevel")
        .and_then(Value::as_str)
        .unwrap_or_default();
    let details = parsed.get("details").cloned().unwrap_or(Value::Null);

    if context == "redis" && level == "warn" {
        return None;
    }

    match message {
        "Starting page" => {
            let page = details
                .get("page")
                .and_then(Value::as_str)
                .unwrap_or("unknown");
            Some(format!("Starting page crawl: {page}"))
        }
        "Page Finished" => {
            let page = details
                .get("page")
                .and_then(Value::as_str)
                .unwrap_or("unknown");
            Some(format!("Page finished: {page}"))
        }
        "Crawl statistics" => {
            let crawled = details.get("crawled").and_then(Value::as_i64).unwrap_or(0);
            let total = details.get("total").and_then(Value::as_i64).unwrap_or(0);
            let pending = details.get("pending").and_then(Value::as_i64).unwrap_or(0);
            let failed = details.get("failed").and_then(Value::as_i64).unwrap_or(0);
            Some(format!(
                "Crawl progress: {crawled}/{total} crawled, {pending} pending, {failed} failed"
            ))
        }
        "Crawling done" => Some("Crawl completed. Starting ZIM conversion...".to_string()),
        _ => {
            if message.is_empty() {
                return None;
            }
            Some(match context {
                "" => message.to_string(),
                "general" | "worker" | "pageStatus" | "crawlStatus" => message.to_string(),
                _ => format!("{context}: {message}"),
            })
        }
    }
}

fn push_recent_output_line(buffer: &mut VecDeque<String>, line: String) {
    let trimmed = line.trim();
    if trimmed.is_empty() {
        return;
    }

    let compact = if trimmed.chars().count() > 280 {
        let shortened = trimmed.chars().take(280).collect::<String>();
        format!("{shortened}...")
    } else {
        trimmed.to_string()
    };

    if buffer
        .back()
        .map(|previous| previous == &compact)
        .unwrap_or(false)
    {
        return;
    }

    if buffer.len() >= 6 {
        buffer.pop_front();
    }
    buffer.push_back(compact);
}

async fn read_log_stream<R>(reader: R, sender: mpsc::UnboundedSender<String>)
where
    R: tokio::io::AsyncRead + Unpin,
{
    let mut lines = BufReader::new(reader).lines();
    loop {
        match lines.next_line().await {
            Ok(Some(line)) => {
                let trimmed = line.trim();
                if trimmed.is_empty() {
                    continue;
                }
                let _ = sender.send(trimmed.to_string());
            }
            Ok(None) => break,
            Err(_) => break,
        }
    }
}

async fn command_succeeds(program: &str, args: &[String]) -> bool {
    match Command::new(program)
        .args(args)
        .stdout(Stdio::null())
        .stderr(Stdio::null())
        .status()
        .await
    {
        Ok(status) => status.success(),
        Err(_) => false,
    }
}

pub async fn check_runtime_health() -> RuntimeHealth {
    let docker_installed = command_succeeds("docker", &["--version".to_string()]).await;

    if !docker_installed {
        return RuntimeHealth {
            docker_installed: false,
            docker_responsive: false,
            zimit_image_present: false,
            ready: false,
            message: Some("Docker is not installed or is not available on PATH.".to_string()),
        };
    }

    let docker_responsive = command_succeeds(
        "docker",
        &[
            "info".to_string(),
            "--format".to_string(),
            "{{.ServerVersion}}".to_string(),
        ],
    )
    .await;

    if !docker_responsive {
        return RuntimeHealth {
            docker_installed: true,
            docker_responsive: false,
            zimit_image_present: false,
            ready: false,
            message: Some("Docker is installed but the daemon is not reachable.".to_string()),
        };
    }

    let zimit_image_present = command_succeeds(
        "docker",
        &[
            "image".to_string(),
            "inspect".to_string(),
            ZIMIT_IMAGE.to_string(),
        ],
    )
    .await;

    let message = if zimit_image_present {
        Some("Runtime ready. ZIM jobs can be started.".to_string())
    } else {
        Some(
            "Docker is ready. Pulling ghcr.io/openzim/zimit may be required on first run."
                .to_string(),
        )
    };

    RuntimeHealth {
        docker_installed: true,
        docker_responsive: true,
        zimit_image_present,
        ready: true,
        message,
    }
}

pub async fn ensure_zimit_image(timeout_duration: Duration) -> Result<bool, String> {
    let inspect_status = tokio::time::timeout(
        Duration::from_secs(20),
        Command::new("docker")
            .args(["image", "inspect", ZIMIT_IMAGE])
            .stdout(Stdio::null())
            .stderr(Stdio::null())
            .status(),
    )
    .await
    .map_err(|_| "Timed out while checking local zimit image state.".to_string())?
    .map_err(|err| format!("Failed to inspect Docker image: {err}"))?;

    if inspect_status.success() {
        return Ok(false);
    }

    let mut pull_cmd = Command::new("docker");
    pull_cmd
        .args(["pull", "--quiet", ZIMIT_IMAGE])
        .stdout(Stdio::null())
        .stderr(Stdio::piped())
        .kill_on_drop(true);

    let pull_child = pull_cmd
        .spawn()
        .map_err(|err| format!("Failed to start Docker image pull: {err}"))?;

    let pull_output = tokio::time::timeout(timeout_duration, pull_child.wait_with_output())
        .await
        .map_err(|_| {
            "Timed out while preparing zimit runtime image. Check Docker connectivity and try again."
                .to_string()
        })?
        .map_err(|err| format!("Failed while waiting for Docker image pull: {err}"))?;

    if pull_output.status.success() {
        Ok(true)
    } else {
        let stderr = String::from_utf8_lossy(&pull_output.stderr);
        Err(format!(
            "Failed to prepare zimit runtime image: {}",
            extract_docker_error(&stderr)
        ))
    }
}

pub fn retry_delay_seconds(attempt_index: u32) -> u64 {
    let exponent = attempt_index.min(4);
    2_u64.pow(exponent)
}

pub fn container_name_for_job(job_id: &str) -> String {
    let suffix = job_id.chars().take(12).collect::<String>();
    format!("zimple-{suffix}")
}

fn docker_mount_path(path: &Path) -> String {
    let raw = path.to_string_lossy().to_string();

    if cfg!(target_os = "windows") {
        let replaced = raw.replace('\\', "/");
        if replaced.len() >= 3 {
            let mut chars = replaced.chars();
            if let (Some(drive), Some(':'), Some('/')) = (chars.next(), chars.next(), chars.next())
            {
                let tail = replaced[3..].to_string();
                return format!("/{}/{tail}", drive.to_ascii_lowercase());
            }
        }
        replaced
    } else {
        raw
    }
}

#[derive(Debug)]
struct DriverScriptGuard {
    path: PathBuf,
}

impl DriverScriptGuard {
    fn prepare(output_dir: &Path, container_name: &str) -> Result<(Self, String), String> {
        let script_filename = format!(".zimple-driver-{container_name}.mjs");
        let script_path = output_dir.join(&script_filename);
        fs::write(&script_path, IDENTITY_ENCODING_DRIVER_SOURCE).map_err(|err| {
            format!(
                "Failed to write zimit compatibility driver at {}: {err}",
                script_path.to_string_lossy()
            )
        })?;

        Ok((
            Self { path: script_path },
            format!("/output/{script_filename}"),
        ))
    }
}

impl Drop for DriverScriptGuard {
    fn drop(&mut self) {
        let _ = fs::remove_file(&self.path);
    }
}

pub fn build_docker_args(
    request: &StartJobRequest,
    output_dir: &Path,
    output_filename: &str,
    container_name: &str,
    driver_container_path: Option<&str>,
) -> Vec<String> {
    let mount = docker_mount_path(output_dir);
    let output_dir_mb = u64::from(request.crawl.limits.max_total_size_mb);
    let size_hard_limit_bytes = output_dir_mb.saturating_mul(1024).saturating_mul(1024);
    let time_hard_limit_seconds =
        u64::from(request.crawl.limits.timeout_minutes).saturating_mul(60);
    let mut args = vec![
        "run".to_string(),
        "--rm".to_string(),
        "--name".to_string(),
        container_name.to_string(),
        "-v".to_string(),
        format!("{mount}:/output"),
        ZIMIT_IMAGE.to_string(),
        "zimit".to_string(),
        "--seeds".to_string(),
        request.url.trim().to_string(),
        "--name".to_string(),
        output_filename.to_string(),
        "--output".to_string(),
        "/output".to_string(),
        "--scopeType".to_string(),
        if request.crawl.include_patterns.is_empty() {
            "domain".to_string()
        } else {
            "custom".to_string()
        },
        "--diskUtilization".to_string(),
        "0".to_string(),
        "-w".to_string(),
        request.crawl.workers.to_string(),
        "--depth".to_string(),
        request.crawl.limits.max_depth.to_string(),
        "--pageLimit".to_string(),
        request.crawl.limits.max_pages.to_string(),
        "--maxPageRetries".to_string(),
        request.crawl.limits.retries.to_string(),
        "--timeHardLimit".to_string(),
        time_hard_limit_seconds.to_string(),
        "--sizeHardLimit".to_string(),
        size_hard_limit_bytes.to_string(),
    ];

    if let Some(driver_path) = driver_container_path {
        args.push("--driver".to_string());
        args.push(driver_path.to_string());
    }

    for include_pattern in &request.crawl.include_patterns {
        args.push("--scopeIncludeRx".to_string());
        args.push(include_pattern.clone());
    }

    for exclude_pattern in &request.crawl.exclude_patterns {
        args.push("--scopeExcludeRx".to_string());
        args.push(exclude_pattern.clone());
    }

    args
}

pub async fn run_zimit_once<F>(
    request: &StartJobRequest,
    output_dir: &Path,
    output_filename: &str,
    container_name: &str,
    mut on_log: F,
) -> Result<DockerRunResult, String>
where
    F: FnMut(String),
{
    let (_driver_guard, driver_container_path) =
        DriverScriptGuard::prepare(output_dir, container_name)?;
    let args = build_docker_args(
        request,
        output_dir,
        output_filename,
        container_name,
        Some(&driver_container_path),
    );

    let mut child = Command::new("docker")
        .args(&args)
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .spawn()
        .map_err(|err| format!("Failed to start Docker process: {err}"))?;

    let stdout = child
        .stdout
        .take()
        .ok_or_else(|| "Failed to capture Docker stdout stream.".to_string())?;
    let stderr = child
        .stderr
        .take()
        .ok_or_else(|| "Failed to capture Docker stderr stream.".to_string())?;

    let (line_tx, mut line_rx) = mpsc::unbounded_channel::<String>();
    let stdout_task = tokio::spawn(read_log_stream(stdout, line_tx.clone()));
    let stderr_task = tokio::spawn(read_log_stream(stderr, line_tx.clone()));
    drop(line_tx);
    let mut recent_output = VecDeque::with_capacity(6);

    while let Some(line) = line_rx.recv().await {
        if !is_docker_noise_line(&line) {
            if let Some(normalized) = normalize_zimit_log_line(&line) {
                push_recent_output_line(&mut recent_output, normalized.clone());
                on_log(normalized);
            }
        }
    }

    let _ = stdout_task.await;
    let _ = stderr_task.await;

    let status = child
        .wait()
        .await
        .map_err(|err| format!("Failed while waiting for Docker process: {err}"))?;

    let success = status.success();
    let error_message = if success {
        None
    } else {
        let mut message = match status.code() {
            Some(code) => format!("zimit container failed with exit code {code}."),
            None => "zimit container terminated by signal.".to_string(),
        };

        if status.code() == Some(2) {
            message.push_str(
                " Exit code 2 usually means zimit rejected one or more options or the output archive name already exists.",
            );
        }

        if !recent_output.is_empty() {
            let context = recent_output
                .iter()
                .cloned()
                .collect::<Vec<String>>()
                .join(" | ");
            message.push_str(" Last output: ");
            message.push_str(&context);
        }

        Some(message)
    };

    Ok(DockerRunResult {
        success,
        error_message,
    })
}

pub async fn stop_container(container_name: &str) -> bool {
    let args = vec![
        "rm".to_string(),
        "-f".to_string(),
        container_name.to_string(),
    ];

    command_succeeds("docker", &args).await
}

pub fn open_path(path: &Path) -> Result<(), String> {
    let mut command = if cfg!(target_os = "macos") {
        let mut cmd = std::process::Command::new("open");
        cmd.arg(path);
        cmd
    } else if cfg!(target_os = "windows") {
        let mut cmd = std::process::Command::new("cmd");
        cmd.arg("/C").arg("start").arg("").arg(path);
        cmd
    } else {
        let mut cmd = std::process::Command::new("xdg-open");
        cmd.arg(path);
        cmd
    };

    command
        .stdout(Stdio::null())
        .stderr(Stdio::null())
        .status()
        .map_err(|err| format!("Unable to open output file: {err}"))
        .and_then(|status| {
            if status.success() {
                Ok(())
            } else {
                Err("Output file open command returned a non-zero status.".to_string())
            }
        })
}

pub async fn sleep_for_retry(attempt_index: u32) {
    let seconds = retry_delay_seconds(attempt_index);
    tokio::time::sleep(Duration::from_secs(seconds)).await;
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::models::{CrawlLimits, CrawlOptions};

    #[test]
    fn retry_backoff_is_bounded() {
        assert_eq!(retry_delay_seconds(0), 1);
        assert_eq!(retry_delay_seconds(1), 2);
        assert_eq!(retry_delay_seconds(4), 16);
        assert_eq!(retry_delay_seconds(12), 16);
    }

    #[test]
    fn container_name_has_prefix() {
        let name = container_name_for_job("1234567890abcdef");
        assert_eq!(name, "zimple-1234567890ab");
    }

    #[test]
    fn builds_expected_arguments() {
        let request = StartJobRequest {
            url: "https://example.com".to_string(),
            output_directory: None,
            output_filename: None,
            crawl: CrawlOptions {
                respect_robots: true,
                workers: 4,
                include_patterns: vec!["/docs".to_string()],
                exclude_patterns: vec!["/admin".to_string()],
                limits: CrawlLimits::default(),
            },
        };

        let args = build_docker_args(
            &request,
            Path::new("/tmp/zimple"),
            "example",
            "zimple-test",
            Some("/output/.zimple-driver-zimple-test.mjs"),
        );

        assert!(args.contains(&"run".to_string()));
        assert!(args.contains(&"zimit".to_string()));
        assert!(args.contains(&"--seeds".to_string()));
        assert!(args.contains(&"https://example.com".to_string()));
        assert!(args.contains(&"--driver".to_string()));
        assert!(args.contains(&"/output/.zimple-driver-zimple-test.mjs".to_string()));
        assert!(args.contains(&"--scopeIncludeRx".to_string()));
        assert!(args.contains(&"--output".to_string()));
        assert!(args.contains(&"/output".to_string()));
    }
}
